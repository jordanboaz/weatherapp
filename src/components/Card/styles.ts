import styled from 'styled-components';

interface ButtonProps {
  active: boolean;
}
export const Container = styled.button<ButtonProps>`
  flex: 1;
  display: flex;
  outline: 0;
  border: 1 solid;
  flex-direction: column;
  align-items: center;
  justify-content: space-evenly;
  border-radius: 1rem;
  padding: 1rem;
  background: ${(props) => props.active ? '#f04a1d' : 'white'};
  color: ${(props) => props.active ? 'white' : '#f04a1d'};
  border-color: var(--orange-500);

  h1 {
      font-weight: bold;
      font-size: 3rem;
    }

  h2 {
    margin-bottom: 1rem;
  }

  h3 {
      margin-top: 1rem;
      font-weight: 600;
    }


  section {
    display: flex;
    width: 100%;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    margin: 0.5rem 0;

   

    
   
    div {
      display: flex;
      flex: 1;
      flex-direction: column;
    }

    h4 {
      font-weight: 400;
      
    }

    p {
      font-weight: 600;
      margin-top: 0.2rem;
    }

   
  }


`
