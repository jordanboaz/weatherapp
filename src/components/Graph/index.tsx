import { ResponsiveBar } from '@nivo/bar'

export function Graph({ data }) {
  return (
    <ResponsiveBar
      data={data}
      keys={["temp"]}
      indexBy="dt"
      margin={{ top: 50, right: 50, bottom: 50, left: 50 }}
      padding={0.3}
      valueScale={{ type: 'linear' }}
      indexScale={{ type: 'band', round: true }}
      colors={'#f04a1d'}
      labelTextColor={"#FFFFFF"}
      axisBottom={{
        tickSize: 5,
        tickPadding: 5,
        tickRotation: 0,
        legend: 'Time',
        legendPosition: 'middle',
        legendOffset: 40
      }}
      axisLeft={{
        tickSize: 5,
        tickPadding: 5,
        tickRotation: 0,
        legend: 'Temperature Â°',
        legendPosition: 'middle',
        legendOffset: -40
      }}
      animate={true}
    />)
}