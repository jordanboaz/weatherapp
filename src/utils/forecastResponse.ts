import { ForecastInfo, Forecast, ForecastWithAggregation } from '../types/weather'

export const restructurePayload = (data: ForecastInfo[]): ForecastWithAggregation[] => {
  const obj = data.reduce((acc, cur) => {

    const datetime = new Date(cur.dt * 1000)

    const date = datetime.toLocaleDateString() // Get the Day
    const time = datetime.toLocaleTimeString() // Get only the time

    acc[date] = { ...acc[date], [time]: cur }

    return acc;
  }, {})

  const aggregated = Object.entries(obj).map(([key, timeChunk]: [key: string, timeChunk: ForecastInfo[]]) => {

    const metrics = calculateMetricsAggregation(Object.values(timeChunk));

    return {
      key,
      metrics,
      timeForecast: timeChunk,
    }

  })

  return aggregated
}

const calculateMetricsAggregation = (data: ForecastInfo[]) => {
  const aggregation = data.reduce((acc, cur) => {
    acc.averageHumidity = acc.averageHumidity + cur.main.humidity;
    acc.averageTemp = acc.averageTemp + cur.main.temp;
    acc.tempMax = Math.max(acc.tempMax, cur.main.temp_max);
    acc.tempMin = Math.min(acc.tempMin, cur.main.temp_min);
    acc.weatherModeAcc.push(cur.weather[0].main)
    return acc;
  }, {
    tempMin: +Infinity,
    tempMax: -Infinity,
    averageHumidity: 0,
    averageTemp: 0,
    weatherModeAcc: [],
    weatherMode: ''
  });

  aggregation.averageHumidity = aggregation.averageHumidity / data.length;
  aggregation.averageTemp = aggregation.averageTemp / data.length;
  aggregation.weatherMode = mode(aggregation.weatherModeAcc)


  return aggregation;

}

const mode = (arr: string[]) => {
  return arr.sort((a, b) =>
    arr.filter(v => v === a).length
    - arr.filter(v => v === b).length
  ).pop();
}
