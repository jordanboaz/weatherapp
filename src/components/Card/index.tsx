import { useDispatch } from 'react-redux';
import { ForecastWithAggregation } from '../../types/weather'
import { Container } from './styles';
import { selectCard } from '../../store/weather'
import { useTypedSelector } from '../../store/useTypedSelector';


interface CardProps {
  data: ForecastWithAggregation
}

const weatherFormatter = new Intl.NumberFormat('de-DE', { maximumSignificantDigits: 3 })
const dateFormatter = new Intl.DateTimeFormat('de-DE', { weekday: 'long', day: 'numeric' })

export function Card({ data }: CardProps) {
  const dispatch = useDispatch();
  const currentCard = useTypedSelector((state) => state.weather?.selectedCard);
  const selected = currentCard?.key;

  const handleClick = () => {
    dispatch(selectCard(data))
  }
  return (
    <Container onClick={handleClick} key={data?.key} active={data?.key === selected} data-testid="cardButton" >
      {data?.key && <h2>{dateFormatter?.format(new Date(data?.key))}</h2>}

      <h1>{weatherFormatter?.format(data?.metrics?.averageTemp)}°</h1>
      <section>
        <div>
          <h4>Min</h4>
          <p>{weatherFormatter?.format(data?.metrics?.tempMin)}°</p>
        </div>
        <div>
          <h4>Max</h4>
          <p>{weatherFormatter?.format(data?.metrics?.tempMax)}°</p>
        </div>
      </section>
      <h3>
        {data?.metrics?.weatherMode}
      </h3>
    </Container>
  )
}