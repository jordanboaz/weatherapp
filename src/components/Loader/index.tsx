import { Container } from './styles'
export function Loader() {
  return (
    <Container>
      <img src="/loader.svg" alt="loading" />
    </Container>
  )
}