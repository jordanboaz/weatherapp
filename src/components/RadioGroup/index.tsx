import { useState } from 'react'
import { Container, RadioItem } from './styles'

type option = {
  key: string;
  value: string
}
interface RadioGroupProps {
  options: option[];
  onChange(key: string): void
}

export function RadioGroup({ options, onChange }: RadioGroupProps) {
  const [selected, setSelect] = useState(0);

  const handleSelect = (index) => {
    setSelect(index)
    onChange(options[index].key)
  }
  return (
    <Container>
      {options.map((option, index) => (
        <RadioItem
          type="button"
          active={index === selected}
          onClick={() => handleSelect(index)}
        >
          {option.value}
        </RadioItem>)
      )}
    </Container>
  )
}