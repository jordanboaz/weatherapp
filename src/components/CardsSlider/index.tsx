import { FiChevronLeft, FiChevronRight } from "react-icons/fi";

import { Container, Arrow } from './styles'
import { Card } from '../Card';
import { useMemo, useState } from 'react';
import { ForecastWithAggregation } from '../../types/weather'

interface CardsSliderProps {
  data: ForecastWithAggregation[];
  cardsToShow: number;
}

const createIndexesArray = (size: number) => Array(size).fill(1).map((v, index) => index)

export function CardsSlider({ data, cardsToShow }: CardsSliderProps) {
  const [indexesShowing, setIndexesShowing] = useState(createIndexesArray(cardsToShow))

  const { first, last } = useMemo(() => {
    return { first: indexesShowing[0], last: indexesShowing[indexesShowing.length - 1] }
  }, [indexesShowing]);


  const { isOnFirst, isOnLast } = useMemo(() => {
    return { isOnFirst: first === 0, isOnLast: last === data.length - 1 }
  }, [first, last])

  const handleLeftArrowClick = () => {
    if (first <= 0) {
      return
    }

    const clone = [...indexesShowing];

    clone.pop();
    clone.unshift(first - 1)
    setIndexesShowing(clone)

  }

  const handleRightArrowClick = () => {

    if (last >= data.length - 1) {
      return
    }

    const clone = [...indexesShowing];

    clone.shift();
    clone.push(last + 1)
    setIndexesShowing(clone)
  }

  console.log(indexesShowing)
  return (
    <Container numCards={cardsToShow}>
      <Arrow onClick={handleLeftArrowClick} data-testid="leftarrow" disabled={isOnFirst}>
        {!isOnFirst && <FiChevronLeft />}
      </Arrow>
      {indexesShowing.map((index) => {
        return <Card key={index} data={data[index]} />
      })}
      <Arrow onClick={handleRightArrowClick} data-testid="rightarrow" disabled={isOnLast} >
        {!isOnLast && <FiChevronRight />}
      </Arrow>
    </Container>
  )
}