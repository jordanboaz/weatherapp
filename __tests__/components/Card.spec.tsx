import React from 'react';
import { Card } from '../../src/components/Card'
import { render } from '../utils/render'

import { Card1, Card2, Cards } from './data'

const initialState = {
  weather: {
    city: undefined,
    count: 0,
    forecast: Cards,
    loading: false,
    units: 'imperial',
    selectedCard: undefined
  }
}

describe("Card", () => {
  it('should render', () => {
    const { asFragment, container } = render(<Card data={Card1 as any} />, {
      initialState
    })

    expect(container).toBeTruthy()
    expect(asFragment()).toMatchSnapshot()
  });

  it('should display infos', () => {
    const { findByText } = render(<Card data={Card1 as any} />, {
      initialState
    })

    expect(findByText(Card1.key)).toBeTruthy()
    expect(findByText(Card1.metrics.tempMin)).toBeTruthy()
    expect(findByText(Card1.metrics.tempMax)).toBeTruthy()
    expect(findByText(Card1.metrics.averageTemp)).toBeTruthy()
    expect(findByText(Card1.metrics.weatherMode)).toBeTruthy()

  })

})