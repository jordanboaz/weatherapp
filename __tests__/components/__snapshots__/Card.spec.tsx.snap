// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render 1`] = `
<DocumentFragment>
  <button
    class="sc-bdvvaa cQLQEp"
    data-testid="cardButton"
  >
    <h2>
      20 Tuesday
    </h2>
    <h1>
      46.1°
    </h1>
    <section>
      <div>
        <h4>
          Min
        </h4>
        <p>
          38.3°
        </p>
      </div>
      <div>
        <h4>
          Max
        </h4>
        <p>
          55.7°
        </p>
      </div>
    </section>
    <h3>
      Clouds
    </h3>
  </button>
</DocumentFragment>
`;
